### PR TITLE
[PM-7066] Use Sensitive in MasterKey

### PR DESCRIPTION
**Pull request by dani-garcia**
## Type of change
```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Use Sensitive in MasterKey and bubble up the Sensitive change for all the requests that contain a password or pin. Note that this requires changing the request structs to be taken by value.

### CODE REVIEW

File: crates/bitwarden-crypto/src/keys/master_key.rs [hunk 1]
```diff
@@ -69,13 +71,17 @@ impl MasterKey {
     }
 
     /// Derives a users master key from their password, email and KDF.
-    pub fn derive(password: &[u8], email: &[u8], kdf: &Kdf) -> Result<Self> {
-        derive_kdf_key(password, email, kdf).map(Self)
+    pub fn derive(password: &SensitiveVec, email: &[u8], kdf: &Kdf) -> Result<Self> {
+        derive_kdf_key(password.expose(), email, kdf).map(Self)
```
Comment (Hinton) [inline crates/bitwarden-crypto/src/keys/master_key.rs:75]: Are we planning on changing `derive_kdf_key` to also accept a sensitive?

File: crates/bitwarden-crypto/src/sensitive/sensitive.rs [hunk 2]
```diff
@@ -67,6 +67,13 @@ impl TryFrom<SensitiveVec> for SensitiveString {
     }
 }
 
+impl From<SensitiveString> for SensitiveVec {
+    fn from(mut s: SensitiveString) -> Self {
+        let value = std::mem::take(&mut s.value);
+        Sensitive::new(Box::new(value.into_bytes()))
```
Comment (Hinton) [inline crates/bitwarden-crypto/src/sensitive/sensitive.rs:73]: Is this getting rid of `value` after the move?

File: crates/bw/src/main.rs [hunk 3]
```diff
@@ -192,10 +193,11 @@ async fn process_commands() -> Result<()> {
 
             let email = text_prompt_when_none("Email", email)?;
             let password = Password::new("Password").prompt()?;
+            let password = SensitiveString::new(Box::new(password));
```
Comment (Hinton) [inline crates/bw/src/main.rs:196]: Should we inline password::new to prevent possibly leaking?

File: crates/bw/src/auth/login.rs [hunk 4]
```diff
@@ -15,12 +16,13 @@ pub(crate) async fn login_password(mut client: Client, email: Option<String>) ->
     let email = text_prompt_when_none("Email", email)?;
 
     let password = Password::new("Password").without_confirmation().prompt()?;
+    let password = SensitiveString::new(Box::new(password));
```
Comment (Hinton) [inline crates/bw/src/auth/login.rs:19]: Inline needed?

File: crates/bitwarden-crypto/src/keys/master_key.rs (hunk 1 above)
Comment (dani-garcia) [reply crates/bitwarden-crypto/src/keys/master_key.rs:75]: derive_kdf_key needs a bit of extra work yet, as the current impl is writing the resulting hash to a fixed size array which is causing leak problems. I'm adding some tests for it now and I'll leave it to a separate PR.

File: crates/bw/src/main.rs (hunk 3 above)
Comment (dani-garcia) [reply crates/bw/src/main.rs]: Done, we'll need to add tests to make sure that the library isn't leaking anything by itself. Not sure how to properly test that yet. Thankfully this isn't used anywhere other than the `bw` crate.

File: crates/bw/src/auth/login.rs (hunk 4 above)
Comment (dani-garcia) [reply crates/bw/src/auth/login.rs]: Resolved with abead06

File: crates/bitwarden-crypto/src/sensitive/sensitive.rs (hunk 2 above)
Comment (dani-garcia) [reply crates/bitwarden-crypto/src/sensitive/sensitive.rs:73]: `into_bytes` consumes the String and returns the internal Vec, so there's no data copied or moved.

**Review by Hinton — APPROVED** [review]

**Pull request merged**